flink-test-utils-parent/flink-connector-testing/src/main/java/org/apache/flink/connectors/test/common/testsuites/SourceTestSuiteBase.java

```diff
@@ -0,0 +1,452 @@
+                                WatermarkStrategy.noWatermarks(),
+                                "Tested Source")
+                        .setParallelism(splitNumber)
+                        .executeAndCollect("Source Multiple Split Test")) {
+            // Check test result
+            checkMultipleSplitRecords(
+                    new MultipleSplitTestData(testRecordCollections), resultIterator);
+        }
+    }
+
+    /**
+     * Test connector source with an idle reader.
+     *
+     * <p>This test will create 4 split in the external system, write test data to all splits, and
+     * consume back via a Flink job with 5 parallelism, so at least one parallelism / source reader
+     * will be idle (assigned with no splits). If the split enumerator of the source doesn't signal
+     * NoMoreSplitsEvent to the idle source reader, the Flink job will never spin to FINISHED state.
+     *
+     * <p>The number and order of records in each split consumed by Flink need to be identical to
+     * the test data written into the external system to pass this test. There's no requirement for
+     * record order across splits.
+     *
+     * <p>A bounded source is required for this test.
+     */
+    @Case
+    @DisplayName("Test source with at least one idle parallelism")
+    public void testIdleReader(TestEnvironment testEnv, ExternalContext<T> externalContext)
+            throws Exception {
+
+        final int splitNumber = 4;
+        final List<Collection<T>> testRecordCollections = new ArrayList<>();
+        for (int i = 0; i < splitNumber; i++) {
+            testRecordCollections.add(generateAndWriteTestData(externalContext));
+        }
+
+        try (CloseableIterator<T> resultIterator =
+                testEnv.createExecutionEnvironment()
+                        .fromSource(
+                                externalContext.createSource(Boundedness.BOUNDED),
+                                WatermarkStrategy.noWatermarks(),
+                                "Tested Source")
+                        .setParallelism(splitNumber + 1)
+                        .executeAndCollect("Redundant Parallelism Test")) {
+            checkMultipleSplitRecords(
+                    new MultipleSplitTestData(testRecordCollections), resultIterator);
+        }
+    }
+
+    /**
+     * Test connector source with task manager failover.
+     *
+     * <p>This test will create 1 split in the external system, write test record set A into the
+     * split, restart task manager to trigger job failover, write test record set B into the split,
+     * and terminate the Flink job finally.
+     *
+     * <p>The number and order of records consumed by Flink should be identical to A before the
+     * failover and B after the failover in order to pass the test.
+     *
+     * <p>An unbounded source is required for this test, since TaskManager failover will be
+     * triggered in the middle of the test.
+     */
+    @Case
+    @Tag("failover")
```

Comment:
constant for tag?

Reply:
I was planning to use this tag to mark cases with failover scenarios, so that user can ignore this cases if their test environment doesn't support failover triggering. 

I'll just remove it for now, and use other annotations to support skipping failover cases in the future.

Reply:
I don't mind that `@Tag` - I was just wondering if we could extract a constant. But I think we are not doing it on other Test tags as well. So just leave.